Refactor deployment

crazy-max opened the pull request:
Closes #14 

* Rename `buildkit-helper` target to `buildkit`
* Merge `publish` job to `build` job in `ci` workflow (now handles with bake-action)
* Review deployment and now using `workflow_dispatch` event with:
  * `qemu_version` input to specify the current QEMU version used
  * `release_number` input to specify the current release number of binfmt
  * `latest` input to create latest tag or not
  * `dry-run` input to release without push
* Change tagging strategy: `<qemu_version>-<release_number>`
* Create following Docker tags on release:
  * `<qemu_version>-<release_number>`, `latest`
  * `buildkit-<qemu_version>-<release_number>`, `buildkit-latest`
* And also create Git tag `<qemu_version>-<release_number>`

tonistiigi commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,72 @@
+        name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      -
+        name: Create Release
```
shouldn't this be after bake?

crazy-max replied:
Yes

crazy-max commented:
@tonistiigi I've changed quite the current behavior:

* Replace `qemu_version` input with `target` to choose between mainline or buildkit bake target.
* Remove `release_number` input and use [`GITHUB_RUN_NUMBER`](https://docs.github.com/en/actions/reference/environment-variables#default-environment-variables) introduced recently.
* I've renamed `QEMU_VERSION` with `QEMU_REF` and resolves `QEMU_VERSION` through [VERSION file](https://github.com/qemu/qemu/blob/v5.2.0/VERSION) which seems more accurate.
* GitHub release body is generated with QEMU version and sha based on QEMU_REF.

tonistiigi commented on Dockerfile:
```diff
@@ -75,9 +76,10 @@ FROM base-$TARGETARCH$TARGETVARIANT AS base
+ARG QEMU_REPO
 RUN --mount=target=.,from=src,src=/src/qemu,rw --mount=target=./install-scripts,src=scripts \
-  TARGETPLATFORM=${TARGETPLATFORM} configure_qemu.sh && \
+  TARGETPLATFORM=${TARGETPLATFORM} QEMU_VERSION=v$(curl -L ${QEMU_REPO}/raw/${QEMU_REF}/VERSION) configure_qemu.sh && \
```
why github specific curl and not just open local file?

crazy-max replied:
right in this case we can use local file

tonistiigi commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,82 @@
+        description: 'Create latest tag'
+        required: true
+        default: 'false'
+      dry-run:
```
would be good to override qemu in here. For test releases, rollbacks etc.

crazy-max replied:
I had to use an HCL function to handle default values. Let me know if it looks good to you: https://github.com/tonistiigi/binfmt/pull/24/commits/7bf70fe0e75c5623d221aafacdfb23d0b4de87e1

tonistiigi commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,82 @@
+          QEMU_VERSION=v$(curl -L ${QEMU_REPO}/raw/${QEMU_REF}/VERSION)
+          QEMU_SHA=$(git ls-remote --sort='version:refname' ${QEMU_REPO} | grep ${QEMU_REF} | cut -f 1 | head -1)
+          TAG=${QEMU_VERSION}-${{ env.GITHUB_RUN_NUMBER }}
+          if [ "${{ github.event.inputs.target }}" = "mainline" ]; then
```
this is opposite I think

crazy-max replied:
yes :)

tonistiigi commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,98 @@
+      -
+        name: Prepare
+        run: |
+          QEMU_REPO=$(docker buildx bake --print ${{ github.event.inputs.target }} | jq --raw-output '.. | .args?.QEMU_REPO | select(.)')
```
why do you parse out the values form bake? instead of just setting the value with a condition like you do already.

crazy-max replied:
I think it's more appropriate to use defaults from the bake def if inputs are empty. This allows us not to duplicate these values in the workflow and also to extract the right values according to the selected target.

tonistiigi replied:
I don't think this is pattern we should be suggesting to any bake users. The logic is that defaults are defined in file and you can override them if needed. Not that you need to use jq to parse out the values from the file and then substitute them with your own values if needed.

crazy-max replied:
I completely agree with that but here we have a rough case where we need to create a GitHub Release in the workflow with its git tag (which is `$QEMU_VERSION-$GITHUB_RUN_NUMBER`) so I have to grab default QEMU_VERSION from bake definition if none is defined as input in the workflow dispatch event.

tonistiigi commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,98 @@
+          if [ "${{ github.event.inputs.qemu_version }}" != "" ]; then
+            QEMU_VERSION=${{ github.event.inputs.qemu_version }}
+          fi
+          QEMU_SHA=$(git ls-remote --sort='version:refname' ${QEMU_REPO} | grep ${QEMU_VERSION} | cut -f 1 | head -1)
```
why is this needed?

crazy-max replied:
Yes it's not needed anymore since I've removed the `qemu_ref` input.

tonistiigi commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,98 @@
+      qemu_repo:
+        description: 'QEMU repo (e.g. https://github.com/qemu/qemu)'
+        required: false
+      qemu_version:
```
how would this do a current buildkit release where source is set with a sha?

crazy-max replied:
`target` input handle the bake target to run. If you set `qemu_version` with a sha the git/docker tag will be the sha commit (maybe we could use a short tag if a long one is provided). That's why I thought it would be better to extract the internal [version available in qemu repo](https://github.com/qemu/qemu/blob/master/VERSION) to create proper docker/git tag.

tonistiigi replied:
Why not just input for the commit tag. That default to source ref (or to not tagging at all)

crazy-max replied:
I guess we could find the most recent tag in the tree from the sha and prepend its value. e.g. `v5.2.0-be25039`, WDYT?

tonistiigi replied:
I don't think we need lots of magic autodetection. This will be used infrequently by a small number of people. They can give the correct inputs, just needs to be a way to enter it.

tonistiigi commented on docker-bake.hcl:
```diff
@@ -21,15 +47,24 @@ target "binaries" {
   args = {
-    QEMU_REPO = "https://github.com/qemu/qemu"
-    QEMU_VERSION = "v5.2.0"
+    QEMU_REPO = trimspace(getdef("${QEMU_REPO}", "https://github.com/qemu/qemu"))
```
This looks super hacky. I guess it is because of the shared var that is updated in build master. But without it I would prefer `--set` then.

crazy-max replied:
> This looks super hacky. I guess it is because of the shared var that is updated in build master.

Yes I would have preferred to use shared var available on current master branch of buildx.

> But without it I would prefer `--set` then.

Yes I will make a change to use `--set`. That's way better.
